hpy/devel/include/hpy/cpython/hpyfunc_trampolines.h

```diff
@@ -75,6 +75,18 @@ typedef HPy (*_HPyCFunction_RICHCMPFUNC)(HPyContext *, HPy, HPy, int);
         return _h2py(func(_HPyGetContext(), _py2h(self), _py2h(obj), op)); \
     }
 
+typedef HPy (*_HPyCFunction_VECTORCALLFUNC)(HPyContext *ctx, HPy callable,
+            HPy const *args, HPy_ssize_t nargsf, HPy kwnames);
+#define _HPyFunc_TRAMPOLINE_HPyFunc_VECTORCALLFUNC(SYM, IMPL)                  \
+    static cpy_PyObject *                                                      \
+    SYM(cpy_PyObject *callable, cpy_PyObject *const *args, size_t nargsf,      \
+            cpy_PyObject *kwnames)                                             \
+    {                                                                          \
+        _HPyCFunction_VECTORCALLFUNC f = (_HPyCFunction_VECTORCALLFUNC)IMPL;   \
+        return _h2py(f(_HPyGetContext(), _py2h(callable),                      \
+                          (HPy *)args, nargsf, _py2h(kwnames)));               \
```

Comment:
> I wonder whether (HPy *)args is a valid C cast or it's an undefined behavior here.

We already do the same cast in `_HPyFunc_TRAMPOLINE_HPyFunc_VARARGS` and it worked so far. I think that the cast is fine as long as the data types have the same size. As you pointed out, `intptr_t` is defined to be large enough to hold the value of a pointer. I agree, there is a caveat: it could be that `sizeof(intptr_t) > sizeof(PyObject *)`. We should probably assert that.

I don't think that a pattern like `(HPy *) (intptr_t *) args` will help because in the end, it's just about how many bytes need to be added to an address and how many bytes need to be read on pointer deref.